File: modules/desktop/home/hyprland/eww/scripts/eww_bar_backend/ctl.py
"""Control client for eww-barctl.

Deliberately isolated from the daemon's import graph. scripts/backend dispatches
here before importing anything else, so a bar click costs one socket round-trip
instead of loading collectors -> urllib.request -> http.client -> email.parser.
Keep this module's imports to json, socket, os and pathlib — the test in
tests/eww_bar_backend/test_ctl.py enforces it.
"""

import json
import os
import socket
from pathlib import Path


CONTROL_USAGE = (
    "usage: eww-barctl ping | volume up|down|set <0-100>|mute|sink <name> | "
    "media play-pause|next|previous | "
    "idle toggle|on|off|status | display normal|external|headless|restore|toggle|status | ai refresh"
    " | bluetooth power-toggle|disconnect <mac> | network wifi-toggle"
    " | notif toggle-group <app>|dismiss <id>|clear-group <app>|clear-all|dnd-toggle|mark-seen"
    " | wallpaper set <path>|rescan"
)


def control_socket_path():
    # Mirrors common.control_socket_path. Duplicated rather than imported so the
    # click path never loads common (and through it subprocess).
    runtime_dir = os.environ.get("XDG_RUNTIME_DIR", "/tmp")
    return Path(runtime_dir) / "eww-backend.sock"


def control_payload_from_args(args):
    if not args or args[0] in ("-h", "--help", "help"):
        raise ValueError(CONTROL_USAGE)
    if args[0] == "ping":
        return {"command": "ping"}
    if args[0] == "volume" and len(args) >= 2:
        payload = {"command": "volume", "action": args[1]}
        if args[1] == "set" and len(args) >= 3:
            payload["value"] = args[2]
        elif args[1] == "sink" and len(args) >= 3:
            payload["sink"] = args[2]
        return payload
    if args[0] == "media" and len(args) == 2:
        return {"command": "media", "action": args[1]}
    if args[0] == "bluetooth" and len(args) >= 2:
        payload = {"command": "bluetooth", "action": args[1]}
        if args[1] == "disconnect" and len(args) >= 3:
            payload["mac"] = args[2]
        return payload
    if args[0] == "network" and len(args) >= 2:
        return {"command": "network", "action": args[1]}
    if args[0] == "idle":
        action = args[1] if len(args) > 1 else "toggle"
        return {"command": "idle", "action": action}
    if args[0] == "display":
        action = args[1] if len(args) > 1 else "status"
        return {"command": "display", "action": action}
    if args[0] == "ai" and len(args) == 2:
        return {"command": "ai", "action": args[1]}
    if args[0] == "notif" and len(args) >= 2:
        payload = {"command": "notif", "action": args[1]}
        if args[1] in ("toggle-group", "clear-group") and len(args) >= 3:
            payload["app"] = " ".join(args[2:])
        elif args[1] == "dismiss" and len(args) >= 3:
            payload["id"] = args[2]
        return payload
    if args[0] == "wallpaper" and len(args) >= 2:
        payload = {"command": "wallpaper", "action": args[1]}
        if args[1] == "set" and len(args) >= 3:
            payload["path"] = args[2]
        return payload
    raise ValueError(CONTROL_USAGE)


def send_control_command(payload):
    with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as client:
        client.connect(str(control_socket_path()))
        client.sendall((json.dumps(payload, separators=(",", ":")) + "\n").encode("utf-8"))
        client.shutdown(socket.SHUT_WR)
        chunks = []
        while True:
            chunk = client.recv(4096)
            if not chunk:
                break
            chunks.append(chunk)
    text = b"".join(chunks).decode("utf-8", errors="replace").strip()
    try:
        return json.loads(text)
    except Exception:
        return {"ok": False, "error": "invalid backend response"}


def run_ctl(args):
    quiet = False
    if args and args[0] in ("-q", "--quiet"):
        quiet = True
        args = args[1:]
    try:
        response = send_control_command(control_payload_from_args(args))
    except Exception as exc:
        response = {"ok": False, "error": str(exc)}
    if not quiet:
        print(json.dumps(response, separators=(",", ":")), flush=True)
    return 0 if response.get("ok") else 1

File: modules/desktop/home/hyprland/eww/client/ctl.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"strings"
)

// controlUsage is kept byte-identical to the Python client's CONTROL_USAGE.
const controlUsage = "usage: eww-barctl ping | volume up|down|set <0-100>|mute|sink <name> | " +
	"media play-pause|next|previous | " +
	"idle toggle|on|off|status | display normal|external|headless|restore|toggle|status | ai refresh" +
	" | bluetooth power-toggle|disconnect <mac> | network wifi-toggle" +
	" | notif toggle-group <app>|dismiss <id>|clear-group <app>|clear-all|dnd-toggle|mark-seen" +
	" | wallpaper set <path>|rescan"

var errUsage = errors.New(controlUsage)

func runtimeFile(name string) string {
	dir := os.Getenv("XDG_RUNTIME_DIR")
	if dir == "" {
		dir = "/tmp"
	}
	return filepath.Join(dir, name)
}

func controlSocketPath() string { return runtimeFile("eww-backend.sock") }

// controlPayloadFromArgs mirrors ctl.control_payload_from_args exactly,
// including which argument counts are accepted and the key order of each
// payload. tests/eww_bar_backend/test_ctl.py pinned this shape; ctl_test.go
// carries the same cases forward.
func controlPayloadFromArgs(args []string) (payload, error) {
	if len(args) == 0 || args[0] == "-h" || args[0] == "--help" || args[0] == "help" {
		return nil, errUsage
	}
	switch args[0] {
	case "ping":
		return payload{{"command", "ping"}}, nil
	case "volume":
		if len(args) >= 2 {
			p := payload{{"command", "volume"}, {"action", args[1]}}
			if args[1] == "set" && len(args) >= 3 {
				p = append(p, field{"value", args[2]})
			} else if args[1] == "sink" && len(args) >= 3 {
				p = append(p, field{"sink", args[2]})
			}
			return p, nil
		}
	case "media":
		if len(args) == 2 {
			return payload{{"command", "media"}, {"action", args[1]}}, nil
		}
	case "bluetooth":
		if len(args) >= 2 {
			p := payload{{"command", "bluetooth"}, {"action", args[1]}}
			if args[1] == "disconnect" && len(args) >= 3 {
				p = append(p, field{"mac", args[2]})
			}
			return p, nil
		}
	case "network":
		if len(args) >= 2 {
			return payload{{"command", "network"}, {"action", args[1]}}, nil
		}
	case "idle":
		action := "toggle"
		if len(args) > 1 {
			action = args[1]
		}
		return payload{{"command", "idle"}, {"action", action}}, nil
	case "display":
		action := "status"
		if len(args) > 1 {
			action = args[1]
		}
		return payload{{"command", "display"}, {"action", action}}, nil
	case "ai":
		if len(args) == 2 {
			return payload{{"command", "ai"}, {"action", args[1]}}, nil
		}
	case "notif":
		if len(args) >= 2 {
			p := payload{{"command", "notif"}, {"action", args[1]}}
			if (args[1] == "toggle-group" || args[1] == "clear-group") && len(args) >= 3 {
				p = append(p, field{"app", strings.Join(args[2:], " ")})
			} else if args[1] == "dismiss" && len(args) >= 3 {
				p = append(p, field{"id", args[2]})
			}
			return p, nil
		}
	case "wallpaper":
		if len(args) >= 2 {
			p := payload{{"command", "wallpaper"}, {"action", args[1]}}
			if args[1] == "set" && len(args) >= 3 {
				p = append(p, field{"path", args[2]})
			}
			return p, nil
		}
	}
	return nil, errUsage
}

// sendControlCommand writes the request and returns the daemon's raw reply.
//
// The reply is returned verbatim rather than re-marshalled: the daemon already
// emits compact, ensure_ascii JSON, so echoing its bytes is byte-identical to
// what the Python client printed after its parse/re-dump round trip -- and it
// cannot reorder keys the way encoding/json would.
func sendControlCommand(p payload) (string, error) {
	conn, err := net.Dial("unix", controlSocketPath())
	if err != nil {
		return "", err
	}
	defer conn.Close()

	if _, err := io.WriteString(conn, p.JSON()+"\n"); err != nil {
		return "", err
	}
	if unixConn, ok := conn.(*net.UnixConn); ok {
		// Mirrors the Python client's shutdown(SHUT_WR): the daemon reads until
		// EOF or newline, and half-closing is what lets it use either.
		if err := unixConn.CloseWrite(); err != nil {
			return "", err
		}
	}
	raw, err := io.ReadAll(conn)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(raw)), nil
}

func runCtl(args []string) int {
	quiet := false
	if len(args) > 0 && (args[0] == "-q" || args[0] == "--quiet") {
		quiet = true
		args = args[1:]
	}

	response, err := func() (string, error) {
		p, err := controlPayloadFromArgs(args)
		if err != nil {
			return "", err
		}
		return sendControlCommand(p)
	}()

	switch {
	case err != nil:
		// Matches the Python client, which wrapped every failure -- bad usage,
		// no socket, a refused connection -- as a failed response rather than a
		// traceback, because these run from :onclick handlers. The `error`
		// text for I/O failures reads differently from CPython's (Go: "dial
		// unix ...: connect: no such file or directory"); the shape, the exit
		// code and the --quiet silence are what eww.yuck depends on, and all 25
		// of its eww-barctl call sites pass --quiet.
		response = errorResponse(err.Error())
	case !json.Valid([]byte(response)):
		response = errorResponse("invalid backend response")
	}

	if !quiet {
		fmt.Println(response)
	}
	if responseIsOK(response) {
		return 0
	}
	return 1
}

func errorResponse(message string) string {
	return `{"ok":false,"error":` + jsonString(message) + `}`
}

func responseIsOK(raw string) bool {
	var decoded map[string]any
	if err := json.Unmarshal([]byte(raw), &decoded); err != nil {
		return false
	}
	ok, _ := decoded["ok"].(bool)
	return ok
}
